Modules/IO/JPEG/src/itkJPEGImageIO.cxx

```diff
@@ -207,30 +206,76 @@ JPEGImageIO::Read(void * buffer)
   // prepare to read the bulk data
   jpeg_start_decompress(&cinfo);
 
-  const auto rowbytes = cinfo.output_width * cinfo.output_components;
-  auto *     tempImage = static_cast<JSAMPLE *>(buffer);
-
   auto * volatile row_pointers = new JSAMPROW[cinfo.output_height];
-  for (size_t ui = 0; ui < cinfo.output_height; ++ui)
+
   {
-    row_pointers[ui] = tempImage + rowbytes * ui;
+    const auto rowbytes = cinfo.output_width * this->GetNumberOfComponents();
+    auto *     tempImage = static_cast<JSAMPLE *>(buffer);
+    for (size_t ui = 0; ui < cinfo.output_height; ++ui)
+    {
+      row_pointers[ui] = tempImage + rowbytes * ui;
+    }
   }
 
   // read the bulk data
-  while (cinfo.output_scanline < cinfo.output_height)
+  if (!(m_IsCMYK && m_CMYKtoRGB))
   {
-    if (setjmp(jerr.setjmp_buffer))
+    while (cinfo.output_scanline < cinfo.output_height)
     {
-      itkWarningMacro("JPEG error in the file " << this->GetFileName());
-      jpeg_destroy_decompress(&cinfo);
-      delete[] row_pointers;
-      return;
+      if (setjmp(jerr.setjmp_buffer))
+      {
+        jpeg_destroy_decompress(&cinfo);
+        delete[] row_pointers;
+        itkWarningMacro(<< "JPEG error in the file " << this->GetFileName());
+        return;
+      }
+
+      jpeg_read_scanlines(&cinfo, &row_pointers[cinfo.output_scanline], cinfo.output_height - cinfo.output_scanline);
     }
-    jpeg_read_scanlines(&cinfo, &row_pointers[cinfo.output_scanline], cinfo.output_height - cinfo.output_scanline);
+
+    // finish the decompression step
+    jpeg_finish_decompress(&cinfo);
   }
+  else
+  {
+    JSAMPROW buf1[1];
+    while (cinfo.output_scanline < cinfo.output_height)
+    {
+      auto * volatile buf0 = new JSAMPLE[cinfo.output_width * 4];
+      buf1[0] = buf0;
+
+      if (setjmp(jerr.setjmp_buffer))
+      {
+        jpeg_destroy_decompress(&cinfo);
+        delete[] row_pointers;
+        delete[] buf0;
+        itkWarningMacro(<< "JPEG error in the file " << this->GetFileName());
+        return;
+      }
+
+      jpeg_read_scanlines(&cinfo, buf1, 1);
+
+      if (cinfo.output_scanline > 0)
+      {
+        const size_t scanline = cinfo.output_scanline - 1;
+        for (size_t i = 0; i < cinfo.output_width; ++i)
+        {
+          // Gimp approach: the following code assumes inverted CMYK values,
+          // even when an APP14 marker doesn't exist. This is the behavior
+          // of recent versions of PhotoShop as well.
+          const float K = buf1[0][4 * i + 3];
+          row_pointers[scanline][3 * i + 0] = static_cast<JSAMPLE>(buf1[0][4 * i + 0] * K / 255.0f);
+          row_pointers[scanline][3 * i + 1] = static_cast<JSAMPLE>(buf1[0][4 * i + 1] * K / 255.0f);
+          row_pointers[scanline][3 * i + 2] = static_cast<JSAMPLE>(buf1[0][4 * i + 2] * K / 255.0f);
+        }
+      }
+
+      delete[] buf0;
```

Comment:
Could this allocation/deallocation be taken out of the loop, so there is only one allocation? If `cinfo.output_width` does not change from scanline to scanline, that should be done.

Reply:
OK.